### DISCUSSION TITLE
Recommended means of developing Baby Buddy on macOS

### DISCUSSION THREAD

**Discussion by skjdghsdjgsdj** [babybuddy/babybuddy#887] [Q&A]
I want to experiment myself with adding [MQTT eventing to Baby Buddy](https://github.com/babybuddy/babybuddy/issues/813). My development environment is macOS with PyCharm. My Mac uses Apple Silicon and thus an arm64 architecture, on which I can also run arm64 VMs. I also have at my disposal a Proxmox host that is amd64 and can run containers and VMs.

How would you suggest setting up a development environment that lets me:

* Edit code in macOS using PyCharm
* _Not_ edit code in a VM (i.e., I don't need to launch PyCharm in the VM or use SSH/Vim/etc. in the VM)
* Best replicate the expected environment that Baby Buddy likes, which seems to be Linux?
* _Not_ run fundamental services like nginx on my Mac directly

What I've done:

* I've followed the [deployment guide](https://docs.baby-buddy.net/setup/deployment/) and got a bare metal version working in a Linux VM on my Mac and also in a Linux container in Proxmox. However, I'd have to edit the source code directly in the VM. I don't want to do that: I want to use PyCharm on my Mac.
* I've used [UTM](https://docs.getutm.app/) on macOS to set up a Debian VM and set up a shared folder that contains all of `babybuddy` (`data` and `public` because why not). The shared folder via VirtFS is atrociously slow to the point of being unusable. Shared folders on Spice WebDAV don't allow for symlinks so the setup fails (`/var/www/babybuddy` is itself a symlink to `/mnt/shared/babybuddy`:
```
/var/www/babybuddy/public$ pipenv install
Creating a virtualenv for this project
Pipfile: /mnt/shared/babybuddy/public/Pipfile
Using default python from /home/.../.local/pipx/venvs/pipenv/bin/python3.11.2 to create virtualenv...
⠋ Creating virtual environment...OSError: [Errno 38] Function not implemented: '/usr/bin/python3.11' -> '/mnt/shared/babybuddy/public/.venv/bin/python'

✘ Failed creating virtual environment
:
Failed to create virtual environment.
```
...so a UTM VM on macOS doesn't seem viable.
* I'm hesitant to share a folder from my Mac to a local VM using the Mac's built-in file sharing services for the security aspects, but also I doubt the protocol I end up picking will support symlinks anyway.
* I want to avoid a bare metal install of Baby Buddy on macOS because I don't want to install nginx or other services on the Mac host. I just want to be able to edit the code from the Mac, but preferably run it from a Linux environment given it seems tailored best to run that way, and it also matches my "production" environment of Baby Buddy too.

So, how do you suggest I develop Baby Buddy on macOS? Is there a way of doing so without this mess of shared folders and VMs but also doesn't require me to install services on my host OS?

**Comment by cdubz** [comment]
> So, how do you suggest I develop Baby Buddy on macOS? Is there a way of doing so without this mess of shared folders and VMs but also doesn't require me to install services on my host OS?

I was recently lured over to VSCode for dev in large part because devcontainers are just so damn easy. I think JetBrains products can use them as well, but when I tried a while back there was some bug preventing it.

If you want to give it a try, you can check out [Dev Containers tutorial](https://code.visualstudio.com/docs/devcontainers/tutorial) -- basically just checkout Baby Buddy in VSCode, reopen the project in the devcontainer, and run `gulp` to start the server in the container and edit code in VSCode.

**Comment by skjdghsdjgsdj** [comment]
I've made progress, kind of, in that you reminded me I don't need nginx and whatnot, just a local dev server.

I made a new PyCharm project by cloning the GitHub repository at `~/Documents/GitHub/babybuddy/public` and made `~/Documents/GitHub/babybuddy/public/media`. I configured the project with a venv at `~/Documents/GitHub/babybuddy/public/.venv` and after activating it ran `pip install -r requirements.txt`. PyCharm also suggested I run `npm install` so I did.

I copied `production.example.py` to `production.py` and created a new `SECRET_KEY`, set `ALLOWED_HOSTS` to `["localhost"]` and for good measure added `DEBUG = True`. That let me successfully run the migrations to set up the SQLite database.

That left me with a mostly automatically generated run configuration which I've modified to be this:

<img width="1016" alt="Screenshot 2024-10-23 at 1 25 40 PM" src="https://github.com/user-attachments/assets/33c00092-1710-4f21-a104-1c67f8290026">

When I run it, my browser at `http://localhost:8000` shows the login and I can log in with the default `admin`/`admin` credentials. But, none of the static assets load: no CSS, images, Javascript, etc. For example:

<img width="584" alt="Screenshot 2024-10-23 at 1 26 58 PM" src="https://github.com/user-attachments/assets/55c1451e-747d-4bc2-ae45-d967eaf7e084">

And `stderr`:
```
[23/Oct/2024 19:20:49] "GET / HTTP/1.1" 302 0
[23/Oct/2024 19:20:49] "GET /login/?next=/ HTTP/1.1" 200 3672
[23/Oct/2024 19:20:49] "GET /static/babybuddy/css/app.css HTTP/1.1" 404 1877
[23/Oct/2024 19:20:49] "GET /static/babybuddy/js/vendor.js HTTP/1.1" 404 1880
[23/Oct/2024 19:20:49] "GET /static/babybuddy/logo/icon-brand.png HTTP/1.1" 404 1901
[23/Oct/2024 19:20:49] "GET /static/babybuddy/js/app.js HTTP/1.1" 404 1871
```

Incidentally, to get the project to launch, I also had to change `site_settings.py` to avoid localized strings or it complained at startup that localized strings weren't yet available:

<img width="1208" alt="Screenshot 2024-10-23 at 1 28 18 PM" src="https://github.com/user-attachments/assets/6cc57dd6-9064-4dc9-9fc5-5b8a9b0749bb">

*edit*: I set a breakpoint to check the value of `STATIC_ROOT` in `base.py` and it shows `~/Documents/GitHub/babybuddy/public/static`.

Any ideas? I feel I'm very close and there's some routing issue, but I don't understand Django enough to know how to troubleshoot it.

**Comment by skjdghsdjgsdj** [comment]
Well, _disabling_ debug by deleting `DEBUG = True` makes the static assets load. I don't understand why as I would hope a debug setting just increases verbosity or something, but apparently it changes static file routing too?

<img width="949" alt="Screenshot 2024-10-23 at 1 48 53 PM" src="https://github.com/user-attachments/assets/8abe4cc9-9ef5-4b0e-9e68-ab33a7df3b10">

**Reply by cdubz** [reply]
@skjdghsdjgsdj yes this has always been a weird and annoying issue. This is where the issue originates: https://github.com/babybuddy/babybuddy/blob/9a89f56067c65781f272265fab0078c5870882fe/babybuddy/urls.py#L60-L61

I don't remember the details exactly, but last time I looked we had to keep that there for some reason...
